x/sync/network_server.go

```diff
@@ -136,32 +145,89 @@ func (s *NetworkServer) HandleChangeProofRequest(
 	}
 
 	// override limit if it is greater than maxKeyValuesLimit
-	limit := req.Limit
-	if limit > maxKeyValuesLimit {
-		limit = maxKeyValuesLimit
+	keyLimit := req.KeyLimit
+	if keyLimit > maxKeyValuesLimit {
+		keyLimit = maxKeyValuesLimit
+	}
+	bytesLimit := int(req.BytesLimit)
+	if bytesLimit > maxByteSizeLimit {
+		bytesLimit = maxByteSizeLimit
 	}
 
-	changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(limit))
-	if err != nil {
-		// handle expected errors so clients cannot cause servers to spam warning logs.
-		if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
-			s.log.Debug(
-				"dropping invalid change proof request",
-				zap.Stringer("nodeID", nodeID),
-				zap.Uint32("requestID", requestID),
-				zap.Stringer("req", req),
-				zap.Error(err),
-			)
-			return nil // dropping request
+	// attempt to get a proof within the bytes limit
+	for keyLimit > 0 {
+		changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(keyLimit))
+		if err != nil {
+			// handle expected errors so clients cannot cause servers to spam warning logs.
+			if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
+				s.log.Debug(
+					"dropping invalid change proof request",
+					zap.Stringer("nodeID", nodeID),
+					zap.Uint32("requestID", requestID),
+					zap.Stringer("req", req),
+					zap.Error(err),
+				)
+				return nil // dropping request
+			}
+			return err
 		}
-		return err
-	}
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
+
+		// ensure that the new limit is always smaller
+		keyLimit = uint16((len(changeProof.KeyValues) + len(changeProof.DeletedKeys)) / 2)
```

Comment:
What's the rationale for doing this instead of just `keyLimit /= 2`?

Reply:
keyLimit/2 might be more keys than were actually returned, which means it will generate another proof that is too large.